File: Cpp/55.cpp
// Given an array of non-negative integers, you are initially positioned at the
// first index of the array.
//
// Each element in the array represents your maximum jump length at that
// position.
//
// Determine if you are able to reach the last index.
//
// Example 1:
//
// Input: [2,3,1,1,4]
// Output: true
// Explanation: Jump 1 step from index 0 to 1, then 3 steps to the last index.
// Example 2:
//
// Input: [3,2,1,0,4]
// Output: false
// Explanation: You will always arrive at index 3 no matter what. Its maximum
//              jump length is 0, which makes it impossible to reach the last
//              index.

#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    bool canJump(vector<int>& nums) {
        int n = nums.size();
        int last = n-1;
        for(int i=n-1;i>=0;--i){
            if(nums[i]+i >= last){
                last = i;
            }
        }
        return last == 0;
    }
};

int main(){
    vector<int> test={3,2,1,0,4};
    Solution s;
    cout<<s.canJump(test)<<endl;

    return 0;
}

File: Python/55.py
class Solution(object):
    def canJump(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        n = len(nums)
        last = n-1
        for i in range(n-1, -1, -1):
            if i+nums[i] >= last:
                last = i
        return last == 0

class Solution(object):
    def canJump(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        if len(nums) <=1:
            return True
        step = 1
        for i in range(len(nums)-2, -1, -1):
            if nums[i] <step:
                step +=1
            else:
                step = 1
        return nums[0] >= step
